File: android/recognition-java/app/src/main/java/eu/id3/face/samples/recognitionjava/FaceProcessor.java
package eu.id3.face.samples.recognitionjava;

import android.content.Context;
import android.hardware.camera2.CameraCharacteristics;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import eu.id3.face.DetectedFace;
import eu.id3.face.DetectedFaceList;
import eu.id3.face.FaceDetector;
import eu.id3.face.FaceEncoder;
import eu.id3.face.FaceException;
import eu.id3.face.FaceLibrary;
import eu.id3.face.FaceMatcher;
import eu.id3.face.FaceModel;
import eu.id3.face.FaceTemplate;
import eu.id3.face.ImageFormat;
import eu.id3.face.Portrait;
import eu.id3.face.PortraitProcessor;
import eu.id3.face.ProcessingUnit;

public class FaceProcessor {
    private FaceDetector faceDetector = null;
    private FaceEncoder faceEncoder = null;

    private PortraitProcessor processor = null;
    private Portrait portrait = null;

    private FaceTemplate enrolledTemplate;

    public FaceProcessor(Context context) {
        String LOG_TAG = "FaceProcessor";
        try {
            /*
             * Load a face detector.
             * First load the model from the Assets and then initialize the FaceDetector object.
             * Only one FaceDetector object is needed to perform all of your detection operation.
             */
            FaceLibrary.loadModelBuffer(
                    readAllBytes(context.getAssets().open("models/face_detector_v4b.id3nn")),
                    FaceModel.FACE_DETECTOR_4B, ProcessingUnit.CPU
            );
            faceDetector = new FaceDetector();
            faceDetector.setConfidenceThreshold(Parameters.detectorConfidenceThreshold);
            faceDetector.setModel(FaceModel.FACE_DETECTOR_4B);
            faceDetector.setThreadCount(Parameters.detectorThreadCount);

            /*
             * Load a face encoder.
             * First load the model from the Assets and then initialize the FaceEncoder object.
             */
            FaceLibrary.loadModelBuffer(
                    readAllBytes(context.getAssets().open("models/face_encoder_v9b.id3nn")),
                    FaceModel.FACE_ENCODER_9B, ProcessingUnit.CPU
            );
            FaceLibrary.loadModelBuffer(
                    readAllBytes(context.getAssets().open("models/face_pose_estimator_v1a.id3nn")),
                    FaceModel.FACE_POSE_ESTIMATOR_1A, ProcessingUnit.CPU
            );
            FaceLibrary.loadModelBuffer(
                    readAllBytes(context.getAssets().open("models/face_occlusion_detector_v2a.id3nn")),
                    FaceModel.FACE_OCCLUSION_DETECTOR_2A, ProcessingUnit.CPU
            );
            FaceLibrary.loadModelBuffer(
                    readAllBytes(context.getAssets().open("models/face_attributes_classifier_v2a.id3nn")),
                    FaceModel.FACE_ATTRIBUTES_CLASSIFIER_2A, ProcessingUnit.CPU
            );
            FaceLibrary.loadModelBuffer(
                    readAllBytes(context.getAssets().open("models/face_landmarks_estimator_v2a.id3nn")),
                    FaceModel.FACE_LANDMARKS_ESTIMATOR_2A, ProcessingUnit.CPU
            );
            faceEncoder = new FaceEncoder();
            faceEncoder.setModel(FaceModel.FACE_ENCODER_9B);
            faceEncoder.setThreadCount(Parameters.encoderThreadCount);

            processor = new PortraitProcessor();

            Log.v(LOG_TAG, "Load models: OK !");
        } catch (FaceException | IOException e) {
            e.printStackTrace();
            Log.e(LOG_TAG, "Error while loading models: " + e.getMessage());
        }
    }

    public static byte[] readAllBytes(InputStream in) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        copyAllBytes(in, out);
        return out.toByteArray();
    }

    /**
     * Copies all available data from in to out without closing any stream.
     */
    public static void copyAllBytes(InputStream in, OutputStream out)
            throws IOException {
        byte[] buffer = new byte[4096];
        while (true) {
            int read = in.read(buffer);
            if (read == -1) {
                break;
            }
            out.write(buffer, 0, read);
        }
    }

    public DetectedFace detectLargestFace(eu.id3.face.Image image) {
        /* Track faces in the image. */
        DetectedFaceList detectedFaceList = faceDetector.detectFaces(image);
        if (detectedFaceList.getCount() > 0) {
            /* At least one face was detected! Return the largest one. */
            return detectedFaceList.getLargestFace();
        } else {
            /* No face was detected. */
            return null;
        }
    }

    EnrollLargestFaceResult enrollLargestFace(eu.id3.face.Image image, DetectedFace detectedFace) {
        /* Create template of the detected face. */
        enrolledTemplate = faceEncoder.createTemplate(image, detectedFace);

        /* Compute template quality to make sure it will good enough for face recognition. */
        int quality = computeQuality(image);

        /* Extracts the portrait image of the detected face to display it. */
        eu.id3.face.Rectangle portraitBounds = detectedFace.getPortraitBounds(0.25f, 0.45f, 1.33f);
        eu.id3.face.Image portraitImage = image.extractRoi(portraitBounds);
        if (Parameters.cameraType == CameraCharacteristics.LENS_FACING_FRONT) {
            portraitImage.flip(true, false);
        }

        /* Compress the portrait image buffer as a JPEG buffer. */
        byte[] jpegPortraitImageBuffer = portraitImage.toBuffer(ImageFormat.JPEG, 75.0f);

        return new EnrollLargestFaceResult(jpegPortraitImageBuffer, quality);
    }

    VerifyLargestFaceResult verifyLargestFace(eu.id3.face.Image image, DetectedFace detectedFace) {
        /* Create template of the detected face. */
        FaceTemplate probeTemplate = faceEncoder.createTemplate(image, detectedFace);

        /* Compute template quality to make sure it will good enough for face recognition. */
        int quality = computeQuality(image);

        /* Initialize a face matcher and compare probe template to the previously enrolled one. */
        FaceMatcher faceMatcher = new FaceMatcher();
        int score = faceMatcher.compareTemplates(probeTemplate, enrolledTemplate);

        return new VerifyLargestFaceResult(quality, score);
    }

    int computeQuality(eu.id3.face.Image image) {
        portrait = processor.createPortrait(image);
        processor.detectOcclusions(portrait);
        processor.estimateFaceAttributes(portrait);
        processor.estimatePhotographicQuality(portrait);

        return portrait.getQualityScore();
    }

    static class EnrollLargestFaceResult {
        private final byte[] jpegPortraitImageBuffer_;
        private final int quality_;

        public EnrollLargestFaceResult(byte[] jpegPortraitImageBuffer, int quality) {
            jpegPortraitImageBuffer_ = jpegPortraitImageBuffer;
            quality_ = quality;
        }

        public byte[] getJpegPortraitImageBuffer() {
            return jpegPortraitImageBuffer_;
        }

        public int getQuality() {
            return quality_;
        }
    }

    class VerifyLargestFaceResult {
        private final int quality_;
        private final int score_;

        public VerifyLargestFaceResult(int quality, int score) {
            score_ = score;
            quality_ = quality;
        }

        public int getScore() {
            return score_;
        }

        public int getQuality() {
            return quality_;
        }
    }
}

File: android/recognition/app/src/main/java/eu/id3/face/samples/recognition/FaceProcessor.kt
package eu.id3.face.samples.recognition

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.media.Image
import android.util.Log
import eu.id3.face.*
import java.nio.ByteBuffer

private const val LOG_TAG = "FaceProcessor"

class FaceProcessor(context: Context) {

    private lateinit var faceDetector: FaceDetector
    private lateinit var faceEncoder: FaceEncoder

    private lateinit var processor: PortraitProcessor
    private lateinit var portrait: Portrait

    private var enrolledTemplate: FaceTemplate? = null

    init {
        try {
            /**
             * Load a face detector.
             * First load the model from the Assets and then initialize the FaceDetector object.
             * Only one FaceDetector object is needed to perform all of your detection operation.
             */
            FaceLibrary.loadModelBuffer(
                context.assets.open("models/face_detector_v4b.id3nn").readBytes(),
                FaceModel.FACE_DETECTOR_4B, ProcessingUnit.CPU
            )
            faceDetector = FaceDetector()
            faceDetector.confidenceThreshold = Parameters.detectorConfidenceThreshold
            faceDetector.model = FaceModel.FACE_DETECTOR_4B
            faceDetector.threadCount = Parameters.detectorThreadCount

            /**
             * Load a face encoder.
             * First load the models from the Assets and then initialize the FaceEncoder object.
             */
            FaceLibrary.loadModelBuffer(
                context.assets.open("models/face_encoder_v9b.id3nn").readBytes(),
                FaceModel.FACE_ENCODER_9B, ProcessingUnit.CPU
            )
            FaceLibrary.loadModelBuffer(
                context.assets.open("models/face_pose_estimator_v1a.id3nn").readBytes(),
                FaceModel.FACE_POSE_ESTIMATOR_1A, ProcessingUnit.CPU
            )
            FaceLibrary.loadModelBuffer(
                context.assets.open("models/face_occlusion_detector_v2a.id3nn").readBytes(),
                FaceModel.FACE_OCCLUSION_DETECTOR_2A, ProcessingUnit.CPU
            )
            FaceLibrary.loadModelBuffer(
                context.assets.open("models/face_attributes_classifier_v2a.id3nn").readBytes(),
                FaceModel.FACE_ATTRIBUTES_CLASSIFIER_2A, ProcessingUnit.CPU
            )
            FaceLibrary.loadModelBuffer(
                context.assets.open("models/face_landmarks_estimator_v2a.id3nn").readBytes(),
                FaceModel.FACE_LANDMARKS_ESTIMATOR_2A, ProcessingUnit.CPU
            )
            faceEncoder = FaceEncoder()
            faceEncoder.model = FaceModel.FACE_ENCODER_9B
            faceEncoder.threadCount = Parameters.encoderThreadCount

            processor = PortraitProcessor()

            Log.v(LOG_TAG, "Load models: OK !")
        } catch (e: FaceException) {
            e.printStackTrace()
            Log.e(LOG_TAG, "Error while loading models: " + e.message)
        }
    }

    fun detectLargestFace(image: eu.id3.face.Image): DetectedFace? {
        /** Track faces in the image. */
        val detectedFaceList = faceDetector.detectFaces(image)
        return if (detectedFaceList.count > 0) {
            /** At least one face was detected! Return the largest one. */
            detectedFaceList.largestFace
        } else {
            /** No face was detected. */
            null
        }
    }

    class EnrollLargestFaceResult(
        private var jpegPortraitImageBuffer: ByteArray,
        private var quality: Int
    ) {
        fun getJpegPortraitImageBuffer(): ByteArray {
            return jpegPortraitImageBuffer
        }

        fun getQuality(): Int {
            return quality
        }
    }

    fun enrollLargestFace(
        image: eu.id3.face.Image,
        detectedFace: DetectedFace
    ): EnrollLargestFaceResult {
        /** Create template of the detected face. */
        enrolledTemplate = faceEncoder.createTemplate(image, detectedFace)

        /** Compute template quality to make sure it will good enough for face recognition. */
        val quality = computeQuality(image)

        /** Extracts the portrait image of the detected face to display it. */
        val portraitBounds = detectedFace.getPortraitBounds(0.25f, 0.45f, 1.33f)
        val portraitImage = image.extractRoi(portraitBounds)
        if (Parameters.cameraType == CameraCharacteristics.LENS_FACING_FRONT) {
            portraitImage.flip(true, false)
        }

        /** Compress the portrait image buffer as a JPEG buffer. */
        val jpegPortraitImageBuffer = portraitImage.toBuffer(ImageFormat.JPEG, 75.0f)

        return EnrollLargestFaceResult(jpegPortraitImageBuffer, quality)
    }

    class VerifyLargestFaceResult(private var quality: Int, private var score: Int) {
        fun getQuality(): Int {
            return quality
        }

        fun getScore(): Int {
            return score
        }
    }

    fun verifyLargestFace(
        image: eu.id3.face.Image,
        detectedFace: DetectedFace
    ): VerifyLargestFaceResult {
        /** Create template of the detected face. */
        val probeTemplate = faceEncoder.createTemplate(image, detectedFace)

        /** Compute template quality to make sure it will good enough for face recognition. */
        val quality = computeQuality(image)

        /** Initialize a face matcher and compare probe template to the previously enrolled one. */
        val faceMatcher = FaceMatcher()
        val score = faceMatcher.compareTemplates(probeTemplate, enrolledTemplate!!)

        return VerifyLargestFaceResult(quality, score)
    }

    fun computeQuality(image: eu.id3.face.Image) : Int {
        portrait = processor.createPortrait(image)
        processor.detectOcclusions(portrait)
        processor.estimateFaceAttributes(portrait)
        processor.estimatePhotographicQuality(portrait)

        return portrait.qualityScore
    }

    companion object {
        fun prepareImageForProcessing(image: Image): eu.id3.face.Image {
            /** Get the YUV planes of the image. */
            val planes: Array<Image.Plane> = image.planes
            val w = image.width
            val h = image.height
            val uvPixelStride: Int = planes[1].pixelStride
            val uvRowStride: Int = planes[1].rowStride
            val plane0 = planes[0].buffer.asReadOnlyBuffer()
            val plane0array = ByteArray(plane0.capacity())
            plane0.get(plane0array)
            val plane1: ByteBuffer = planes[1].buffer.asReadOnlyBuffer()
            val plane1array = ByteArray(plane1.capacity())
            plane1.get(plane1array)
            val plane2: ByteBuffer = planes[2].buffer.asReadOnlyBuffer()
            val plane2array = ByteArray(plane2.capacity())
            plane2.get(plane2array)

            /** Convert this image to an id3.face Image. */
            return eu.id3.face.Image.fromYuvPlanes(
                plane0array,
                plane1array,
                plane2array,
                w,
                h,
                uvPixelStride,
                uvRowStride,
                PixelFormat.BGR_24_BITS
            )
        }
    }
}